connmgr/conngating.go

```diff
@@ -0,0 +1,43 @@
+package connmgr
+
+import (
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// AddressConnectionGater can be implemented by a type that supports active
+// connection gating based on the multi-address of the peer.
+type AddressConnectionGater interface {
+	// DenyAddrConnection returns true if a connection to/from a peer
+	// should be denied.
+	// The caller must supply the multi-address of the remote peer.
+	DenyAddrConnection(ma.Multiaddr) (deny bool)
```

Comment:
Is it worth providing both `(ma.Multiaddr, peer.ID)` as arguments. Rather than re-parsing the peer out of the address, most dial attempts ought to be in the context of a specific peer already (so it shouldn't be much of an imposition on the caller), and that might make implementing some type of gaters simpler.